docs/dag-sharding-rfc.md

```diff
@@ -0,0 +1,176 @@
+# Introduction
+This document is a rough draft of an RFC for supporting sharded dags in ipfs-cluster.  After these ideas become more concrete, a little more thoroughly researched and have had time to incorporate changes suggested by ipfs-cluster contributors the goal is to include this RFC in ipfs/notes to get comments from the rest of the ipfs community.
+
+This document outlines motivations for sharding dags across the nodes of an ipfs-cluster, and provides suggestions on development paths to achieve the motivated goals.  It begins by making the strongest assumptions about the data, starting with familiar unixfs dags describing ipfs files that can be stored on any single node's repo, and working up to a discussion of arbitrary ipld dags that are too large to fit on any node's repo.
+
+# Motivation
+There are two primary motivations for adding data sharding to ipfs-cluster.  See the WIP use cases documents in PR #215, issue #212, and posts on the discuss.ipfs forum (ex: https://discuss.ipfs.io/t/ubuntu-archive-on-top-of-ipfs/1579 and https://discuss.ipfs.io/t/segmentation-file-in-ipfs-cluster/1465/2 and https://discuss.ipfs.io/t/when-file-upload-to-node-that-can-ipfs-segmentation-file-to-many-nodes/1451) for some more context.  (TODO -- investigate b5/data-together use cases and document whether this shows up there, if so add to use case doc + link here)
+
+## Motivation 1: store files too big for a single node
+This is one of the most requested features for ipfs-cluster.  ipfs-pack (https://github.com/ipfs/ipfs-pack) exists to fill a similar need by allowing data stored in a POSIX filesystem to be referenced from an ipfs-node's datastore without acutally copying all of the data into the node's repo.  However certain use cases require functionality beyond ipfs-pack. For example, what if a cluster of nodes needs to download a large file bigger than any of its nodes' individual machines' disk size?  In this case using ipfs-pack is not enough, such massive files would not fit on a single node's storage device and coordination between nodes to store the file becomes essential.
+
+Storing large files is a feature that would serve many use cases, the more general ability to store large ipld dags of arbitrary structure has the potential to support others.  As an example consider importing a cryptocurrency ledger to an ipfs cluster when no one node could hold the entire dag.
+
+## Motivation 2: store files with different replication strategies for fault tolerance
+This feature is also frequently requested, including in the original thread leading to the creation of ipfs-cluster (See issue #1), and in @raptortech-js 's thorough discussion in issue #9.  The idea is to bring RAID modes of storage, including modes that incorporate space-efficient FEC techniques, to provide better fault tolerance of storage.  This would open up ipfs-cluster to new usage patterns, especially storing archives with infrequent updates and long storage periods.
```

Comment:
I understand your perspective.  My reason for including RAID specifically stems from the following passage in the original cluster design proposal linked in issue #1 (https://github.com/ipfs/notes/issues/58):
> **We want**
>...
>*(Mostly Certain) Use RAID configurations for "pin-replication" cluster modes. 
>  * RAID modes are well understood and well documented. so they're the obvious default choice.
>  * but we should consider other well understood data replication configuration systems.

It would be helpful at some point to go through this proposal and issue #1 together so I can know for sure what is still relevant.

Reply:
Yeah I know where this comes from, but I think talking about RAID modes does not make understanding easier. You could say cluster now supports RAID 0 (replication level 1) and RAID 1 (replication level = 2). Is a larger replication factor a different RAID level?

It does not make sense to speak of byte-level stripping in cluster, so would are we talking about ipfs-block-level stripping instead? Or merkle-tree stripping? Would ipfs-block-level stripping make sense? For performance? Is it better that ipfs-blocks from one file are in the same daemon or potentially is as many daemons as possible? Most RAID modes are not really used (according to wikipedia), do we aim to support them?. How does the RAID 6 standard deal with disks being added or removed? 

Sorry but I find it super confusing to speak in terms of RAID vs:

* Your data is replicated X times in different nodes
* Or, your tree is distributed in different nodes and includes a checksum for recovery in case of disaster
* Or, each branch of your tree is stored in a different node

I don't even know how much of the RAID motivations and error correction codes are part of spinning magnetic disk scenarios (which have certain performance issues, and certain types of errors). So yep, I don't feel comfortable expressing our capabilities in terms of RAID levels.

Reply:
@hsanjuan solid points all around.  Thank you for elaborating.